File: Leetcode100/leetcode100_python/BestTimetoBuyandSellStockwithCooldown2.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        n = len(prices)
        dp = [[0 for _ in range(3)] for __ in range(n)]
        for i in range(n):
            if i == 0:
                dp[i][0] = -prices[i]
            else:
                dp[i][0] = max(dp[i - 1][0], dp[i - 1][2] - prices[i])
                dp[i][1] = dp[i - 1][0] + prices[i]
                dp[i][2] = max(dp[i - 1][1], dp[i - 1][2])
        return max(dp[-1][1], dp[-1][2])


if __name__ == "__main__":
    prices = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.maxProfit(prices))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/best_time_to_buy_and_sell_stock_with_cooldown/BestTimetoBuyandSellStockwithCooldown2.java
package main.java.com.jiang.leetcode100.best_time_to_buy_and_sell_stock_with_cooldown;

import java.util.Scanner;

class Solution2 {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if(n == 0) return 0;

        int[][] dp = new int[n][3];  // dp[i][0]:手上持有股票 dp[i][1]:不持有，冷冻期 dp[i][2] 不持有，非冷冻期

        for(int i = 0; i < n; i++) {
            if(i == 0) {
                dp[i][0] = -prices[i];
            } else {
                dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][2] - prices[i]);
                dp[i][1] = dp[i - 1][0] + prices[i];
                dp[i][2] = Math.max(dp[i - 1][1], dp[i - 1][2]);
            }
        }
        return Math.max(dp[n - 1][1], dp[n - 1][2]);
    }
}

public class BestTimetoBuyandSellStockwithCooldown2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] prices = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) prices[i] = Integer.parseInt(nums_str[i]);

        System.out.println(new Solution2().maxProfit(prices));
    }
}